# Práctica 1

## Descripción

**Problema**: Crear un modelo de clasificación que prediga si va a llover o no mañana.

**Data set**: 145.000 observaciones diarias con 24 atributos. Usaremos 23 de ellas para predecir si llueve o no.

**Lógica**: Construir un modelo de clasificación que prediga si va a llover o no.

## Código

### Cargar librerías

1) Cargar librerías *numpy* y *pandas*.

In [161]:
import numpy as np
import pandas as pds

### Cargar datos

2) Cargar el dataset weatherAUS.csv usando el método *read_csv* de pandas.

In [162]:
weatherPath = './weatherAUS.csv'
pds.read_csv(weatherPath)

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  \
0       2008-12-01   Albury     13.4     22.9       0.6          NaN   
1       2008-12-02   Albury      7.4     25.1       0.0          NaN   
2       2008-12-03   Albury     12.9     25.7       0.0          NaN   
3       2008-12-04   Albury      9.2     28.0       0.0          NaN   
4       2008-12-05   Albury     17.5     32.3       1.0          NaN   
...            ...      ...      ...      ...       ...          ...   
142188  2017-06-20    Uluru      3.5     21.8       0.0          NaN   
142189  2017-06-21    Uluru      2.8     23.4       0.0          NaN   
142190  2017-06-22    Uluru      3.6     25.3       0.0          NaN   
142191  2017-06-23    Uluru      5.4     26.9       0.0          NaN   
142192  2017-06-24    Uluru      7.8     27.0       0.0          NaN   

        Sunshine WindGustDir  WindGustSpeed WindDir9am  ... Humidity3pm  \
0            NaN           W           44.0          W  ...        22.0   
1            NaN         WNW           44.0        NNW  ...        25.0   
2            NaN         WSW           46.0          W  ...        30.0   
3            NaN          NE           24.0         SE  ...        16.0   
4            NaN           W           41.0        ENE  ...        33.0   
...          ...         ...            ...        ...  ...         ...   
142188       NaN           E           31.0        ESE  ...        27.0   
142189       NaN           E           31.0         SE  ...        24.0   
142190       NaN         NNW           22.0         SE  ...        21.0   
142191       NaN           N           37.0         SE  ...        24.0   
142192       NaN          SE           28.0        SSE  ...        24.0   

        Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  \
0            1007.7       1007.1       8.0       NaN     16.9     21.8   
1            1010.6       1007.8       NaN       NaN     17.2     24.3   
2            1007.6       1008.7       NaN       2.0     21.0     23.2   
3            1017.6       1012.8       NaN       NaN     18.1     26.5   
4            1010.8       1006.0       7.0       8.0     17.8     29.7   
...             ...          ...       ...       ...      ...      ...   
142188       1024.7       1021.2       NaN       NaN      9.4     20.9   
142189       1024.6       1020.3       NaN       NaN     10.1     22.4   
142190       1023.5       1019.1       NaN       NaN     10.9     24.5   
142191       1021.0       1016.8       NaN       NaN     12.5     26.1   
142192       1019.4       1016.5       3.0       2.0     15.1     26.0   

        RainToday  RISK_MM  RainTomorrow  
0              No      0.0            No  
1              No      0.0            No  
2              No      0.0            No  
3              No      1.0            No  
4              No      0.2            No  
...           ...      ...           ...  
142188         No      0.0            No  
142189         No      0.0            No  
142190         No      0.0            No  
142191         No      0.0            No  
142192         No      0.0            No  

[142193 rows x 24 columns]

3) Mostrar el tamaño del data set usando el método *shape* de pandas.

In [163]:
df = pds.read_csv(weatherPath)
df10 = pds.read_csv(weatherPath)
df.shape

(142193, 24)

4) Mostrar las 5 primeras líneas del data set usando cualquier método.

In [164]:
print(df[0:5])

         Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1  2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2  2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3  2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4  2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity3pm  Pressure9am  \
0           W           44.0          W  ...        22.0       1007.7   
1         WNW           44.0        NNW  ...        25.0       1010.6   
2         WSW           46.0          W  ...        30.0       1007.6   
3          NE           24.0         SE  ...        16.0       1017.6   
4           W           41.0        ENE  ...        33.0       1010.8   

   Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  RISK_MM  \
0       1007.1       8

### Pre-procesado

5) En todos los data sets puede haber atributos sesgados, incorrectos, imprecisos etc. Una forma de comprobar qué atributos son es contar los valores y compararlo con el tamaño obtenido en 3). Para ello utilizaremos el método *count* y el método *sort_values* de pandas.

In [165]:
# df = df.sort_values(by='')
# print(df.count())
# df = data.sort_values(by=count)
# print(df)

6) Es necesario borrar todas las variables que no sean significativas. Para ello borraremos las siguientes: 'Sunshine','Evaporation','Cloud3pm','Cloud9am','Location','RISK_MM','Date'. RISK_MM denota la cantidad de lluvia prevista al día siguiente y puede dañar nuestro modelo.

Para borrar columnas el método *drop* de pandas, y mostraremos a continuación el estado del data set por medio del método *shape*.

In [166]:
df = df.drop(columns=['Sunshine', 'Evaporation', 'Cloud3pm', 'Cloud9am', 'Location', 'RISK_MM'])
df.shape
print(df)

              Date  MinTemp  MaxTemp  Rainfall WindGustDir  WindGustSpeed  \
0       2008-12-01     13.4     22.9       0.6           W           44.0   
1       2008-12-02      7.4     25.1       0.0         WNW           44.0   
2       2008-12-03     12.9     25.7       0.0         WSW           46.0   
3       2008-12-04      9.2     28.0       0.0          NE           24.0   
4       2008-12-05     17.5     32.3       1.0           W           41.0   
...            ...      ...      ...       ...         ...            ...   
142188  2017-06-20      3.5     21.8       0.0           E           31.0   
142189  2017-06-21      2.8     23.4       0.0           E           31.0   
142190  2017-06-22      3.6     25.3       0.0         NNW           22.0   
142191  2017-06-23      5.4     26.9       0.0           N           37.0   
142192  2017-06-24      7.8     27.0       0.0          SE           28.0   

       WindDir9am WindDir3pm  WindSpeed9am  WindSpeed3pm  Humidity9am  \
0 

7) Ahora eliminaremos los datos nulos utilizando para ello el método *dropna* con el parámetro *how='any'* [info](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html).
  Después, mostraremos el estado del data set por medio del método *shape*.

In [167]:
df = data.dropna(how='any')
df.shape
print(df)

              Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  \
5939    2009-01-01    Cobar     17.9     35.2       0.0         12.0   
5940    2009-01-02    Cobar     18.4     28.9       0.0         14.8   
5942    2009-01-04    Cobar     19.4     37.6       0.0         10.8   
5943    2009-01-05    Cobar     21.9     38.4       0.0         11.4   
5944    2009-01-06    Cobar     24.2     41.0       0.0         11.2   
...            ...      ...      ...      ...       ...          ...   
139108  2017-06-20   Darwin     19.3     33.4       0.0          6.0   
139109  2017-06-21   Darwin     21.2     32.6       0.0          7.6   
139110  2017-06-22   Darwin     20.7     32.8       0.0          5.6   
139111  2017-06-23   Darwin     19.5     31.8       0.0          6.2   
139112  2017-06-24   Darwin     20.2     31.7       0.0          5.6   

        Sunshine WindGustDir  WindGustSpeed WindDir9am  ... Humidity3pm  \
5939        12.3         SSW           48.0        ENE  ... 

8) Es interesante comprobar si nuestros datos tienen *outliers* (valroes atípicos), estos son datos que difieren significativamente de otros. Los *outliers* son consecuencia de cálculos erróneos cuando se recopilan los datos.

Para esto vamos a emplear el módulo *stats* de la librería SciPy (herramientas y algoritmos matemáticos)

    from scipy import stats
    
A continuación obtenemos el valor absoluto de la desviación típica de todos los datos.

    z = np.abs(stats.zscore(df._get_numeric_data()))

Un valor de z-score mayor que 3 nos dice que el valor es mucho mayor que el promedio, un valor igual a 0 que está en el promedio.

[Standard Normal Distribution](http://www.ltcconline.net/greenl/courses/201/probdist/ctsdis8.jpg)


Y utilizamos ese valor para mostrar solo aquellos datos que sean menores que 3 en base a las columnas (axis = 1).

    df= df[(z < 3).all(axis=1)]
    
Por último, mostramos el estado del data set con *shape*.

In [168]:
from scipy import stats

z = np.abs(stats.zscore(df._get_numeric_data()))
df= df[(z < 3).all(axis=1)]
df.shape

(52515, 24)

9) Vamos a emplear 0 y 1 en lugar de  *labels* (etiquetas), en los atributos RainToday y RainTomorrow. Utiliza el método *replace* de pandas.

In [169]:
df["RainToday"].replace({"No": 0}, inplace=True)
df["RainTomorrow"].replace({"No": 1}, inplace=True)
print(df)

              Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  \
5939    2009-01-01    Cobar     17.9     35.2       0.0         12.0   
5940    2009-01-02    Cobar     18.4     28.9       0.0         14.8   
5942    2009-01-04    Cobar     19.4     37.6       0.0         10.8   
5943    2009-01-05    Cobar     21.9     38.4       0.0         11.4   
5944    2009-01-06    Cobar     24.2     41.0       0.0         11.2   
...            ...      ...      ...      ...       ...          ...   
139108  2017-06-20   Darwin     19.3     33.4       0.0          6.0   
139109  2017-06-21   Darwin     21.2     32.6       0.0          7.6   
139110  2017-06-22   Darwin     20.7     32.8       0.0          5.6   
139111  2017-06-23   Darwin     19.5     31.8       0.0          6.2   
139112  2017-06-24   Darwin     20.2     31.7       0.0          5.6   

        Sunshine WindGustDir  WindGustSpeed WindDir9am  ... Humidity3pm  \
5939        12.3         SSW           48.0        ENE  ... 

10) Normalizaremos los datos para que no haya distorsiones entre las diferencias del rango de valores. Esto es necesario cuando las variables tienen diferentes rangos. ¿Por qué? Veamos un ejemplo: 

Tenemos los atributos edad (0-100) y salario (0-100.000), al hacer una un análisis como una regresión lineal múltiple, el atributo salario tendrá más influencia debido a su mayor valor.

Para llevar a cabo la normalización vamos a emplear la función *MinMaxScaler* del módulo *preprocessing* de la librería *sklearn*. (sklearn es como se denota scikit-learn en Python)

    from sklearn import preprocessing
    scaler = preprocessing.MinMaxScaler()
    scaler.fit(df)

Mostraremos algunos datos de nuestro data set.
(Esta función nos dará error ¿Por qué? ¿Cómo solucionarlo?)

In [170]:
from sklearn import preprocessing

arreglo = preprocessing.LabelEncoder()
df = df10.apply(arreglo.fit_transform)

scaler = preprocessing.MinMaxScaler()
scaler.fit(df)

df = pds.DataFrame(scaler.transform(df), index=df.index, columns=df.columns)
df.iloc[4:10]

TypeError: Encoders require their input to be uniformly strings or numbers. Got ['float', 'str']

### Análisis exploratorio de datos (Exploratory Data Analysis - EDA)

11) Vamos a identificar las variables más significativas para predecir nuestra función. Para ello utilizaremos la función *SelectKBest* [info](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.SelectKBest.html) del sklearn. Utilizaremos el método predictivo chi2 para puntuar a cada una de ellas y seleccionaremos solo 3.

    from sklearn.feature_selection import SelectKBest, chi2
    X = df.loc[:,df.columns!='RainTomorrow']
    y = df[['RainTomorrow']]
    selector = SelectKBest(chi2, k=3)
    selector.fit(X, y)
    
Y seleccionamos las variables con *get_support*.

    print(X.columns[selector.get_support(indices=True)])

'Rainfall', 'Humidity3pm' y 'RainToday' son las variables más significativas.

### Construyendo un modelo de aprendizaje automático

12) En primer lugar, y para simplificar el modelo, vamos a emplear solo 'Humidity3pm' para construirlo. De esta forma tendremos 'Humidity3pm' como entrada y 'RainTomorrow' como salida. 
Crea una variable 'X' formada únicamente por la columna del *dataframe* 'Humidity3pm' y otra variable 'y' formada por la columna del *dataframe* 'RainTomorrow'.

13) Vamos a construir modelos de clasificación utilizando los distintos algoritmos:
- *Logistic regression* (Regresión logística) [info](https://en.wikipedia.org/wiki/Logistic_regression) 
- *Random Forest* (bosques aleatorios) [info](https://en.wikipedia.org/wiki/Random_forest) 
- *Decission Tree* (Árbol de decisión) [info](https://en.wikipedia.org/wiki/Decision_tree) 
- *Support Vector Machine* (Máquinas de vectores de soporte) [info](https://en.wikipedia.org/wiki/Support_vector_machine) 


No te preocupes si todavía no los conoces, profundizaremos sobre ellos a lo largo del curso.

#### 13.1) Logistic regression
    

- Cargamos distintos recursos la librería sklearn. Concretamente necesitaremos el modelo de *logistic regression*, un método para obtener automáticamente los datos de entrenamiento y test, y otro método para obtener la precisión del modelo. 


    from sklearn.linear_model import LogisticRegression
    from sklearn.model_selection import train_test_split
    from sklearn.metrics import accuracy_score

- Dividimos los datos que tenemos en datos de entrenamiento o *training set* (conjunto de datos iniciales utilizados por el modelo para entender el problema y aprender sobre él) y datos de prueba o *test set* (conjunto de datos utilizados para comprobar lo que ha aprendido el modelo después de haber sido entrenado). __Nunca uses los datos de prueba para el entrenamiento o tus resultados estarían comprometidos__. Normalmente se utilzia un 80% de los datos para entrenamiento y un 20% para tests.


    X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.20)

- Construimos el modelo y lo entrenamos con el *training set*


    clf_logreg = LogisticRegression(random_state=0)
    clf_logreg.fit(X_train,y_train)

(Ignora el *warning*)

- Evaluamos el modelo usando el *test set*.

    y_pred = clf_logreg.predict(X_test)
    score = accuracy_score(y_test,y_pred)

- Por último mostramos por pantalla el *score* del modelo en %.

#### 13.2) Random Forest

- Crea un modelo Random Forest [info](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html), entrénalo y evalúa el modelo siguiendo los pasos descritos en el punto 13.1)


    from sklearn.ensemble import RandomForestClassifier
    

- Cuando construyas el modelo, utiliza los siguientes parámetros:


    n_estimators=100, max_depth=4,random_state=0

(Ignora el *warning*)    

#### 13.3) Decision Tree

- Crea un modelo Decision Tree [info](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html), entrénalo y evalúa el modelo siguiendo los pasos descritos en el punto 13.1)


    from sklearn.tree import DecisionTreeClassifier
    

- Cuando construyas el modelo, utiliza los siguientes parámetros:


    random_state=0    
    
(Ignora el *warning*)

#### 13.4) Support Vector Machine

Crea un modelo SVM, entrénalo y evalúa el modelo siguiendo los pasos descritos en el punto 13.1)
    from sklearn import svm
- Crea un modelo SVM [info](https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html), entrénalo y evalúa el modelo siguiendo los pasos descritos en el punto 13.1)


    from sklearn import svm
    

- Cuando construyas el modelo, utiliza los siguientes parámetros:


    kernel='linear'
    
(Este modelo tarda en torno a 1 minuto en ejecutarse)
(Ignora el *warning*)

### Final

14) Ordena de mayor a menor precisión los modelos.

15) ¿Con qué término en inglés se denota el sobreentrenamiento de un modelo de ML y qué características tiene?